### PR TITLE
Reverse proxy support

### PR DESCRIPTION
**Pull request by Altai-man**

### CODE REVIEW

File: lib/Cro/HTTP/ResponseParser.pm6 [hunk 1]
```diff
@@ -51,7 +51,7 @@ class Cro::HTTP::ResponseParser does Cro::Transform {
 
                         # Validate version.
                         my $version = $parsed[0].Str;
-                        unless $version.starts-with('1') {
```
Comment (jnthn) [inline lib/Cro/HTTP/ResponseParser.pm6:54]: Why is this needed?

File: lib/Cro/HTTP/Request.pm6 [hunk 2]
```diff
@@ -83,6 +83,7 @@ class Cro::HTTP::Request does Cro::HTTP::Message {
         die X::Cro::HTTP::Request::Incomplete.new(:missing<method>) unless $!method;
         die X::Cro::HTTP::Request::Incomplete.new(:missing<target>) unless $!target;
         my $version = self.http-version // (self.has-header('Host') ?? '1.1' !! '1.0');
+        $version = '1.1' if $version eq '2.0';
```
Comment (jnthn) [inline lib/Cro/HTTP/Request.pm6:86]: Hmm, wonder if better would be to add a `:$force-version` named parameter, so we don't lose the "debug dump" value of it showing 2.0 when that's where it originally came from?

File: lib/Cro/HTTP/Response.pm6 [hunk 3]
```diff
@@ -82,7 +82,9 @@ class Cro::HTTP::Response does Cro::HTTP::Message {
         my $status = $!status // (self.has-body ?? 200 !! 204);
         my $reason = %reason-phrases{$status} // 'Unknown';
         my $headers = self!headers-str();
-        "HTTP/{self.http-version // '1.1'} $status $reason\r\n$headers\r\n"
+        my $ver = self.http-version // '1.1';
+        $ver = '1.1' if $ver eq '2.0';
+        "HTTP/$ver $status $reason\r\n$headers\r\n"
```
Comment (jnthn) [inline lib/Cro/HTTP/Response.pm6:87]: And perhaps `:$force-version` here too

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 4]
```diff
@@ -0,0 +1,91 @@
+use Cro;
+use Cro::HTTP::Client;
+
+class Cro::HTTP::ReverseProxy does Cro::Transform {
+    has $.to;
+    has $.to-absolute;
+    has $!destination;
+    has $!client;
+    has $.request;
+    has $.response;
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:10]: Maybe type these?

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    has %.ca;
+
+    submethod TWEAK() {
+        unless ($!to.defined ^^ $!to-absolute.defined) {
+            die 'Either `to` or `to-absolute` must be specified for ReverseProxy';
+        }
+        $!to = self!trim-url($!to) if $!to ~~ Str;
+        $!to-absolute = self!trim-url($!to-absolute) if $!to-absolute ~~ Str;
+        with $!to {
+            die '`to` of ReverseProxy must be a Str or code block' unless $!to ~~ Str|Callable;
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:20]: Perhaps nicer as a subset type?

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    method consumes() { Cro::HTTP::Request }
+    method produces() { Cro::HTTP::Response }
+
+    method !trim-url($url) { $url.ends-with('/') ?? $url.substr(0, *-1) !! $url }
+
+    method transformer(Supply $pipeline --> Supply) {
+        supply whenever $pipeline -> $request {
+            my $*PROXY-STATE;
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:38]: I think we'll have to declare a leixcal `my $proxy-state` and this level, and bind it like `my $*PROXY-STATE := $proxy-state;` ahead of calls that should be able to see it in order to get it robustly available even over async boundaries.

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            sub send($request) {
+                my %options = headers => $request.headers,
+                              body => await $request.body if $request.has-body;
+
+                whenever self!make-destination($request) {
+                    my $response = await $!client.request($request.method, $_, %options);
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:44]: We're mixing `whenever` and `await` here and it's not entirely clear to me we want to be. `await` blocks the request pipeline, so I think if we had other HTTP/2 streams we'd not be able to concurrently handle the proxying of the requests so efficiently with the `await`.

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    }
+
+    method !make-destination(Cro::HTTP::Request $request) {
+        my $target = $request.target;
+        $target .= substr(1..*) if $target.starts-with('/');
+        if $!destination ~~ Str {
+            my $p = Promise.new;
+            $p.keep($!to ?? "{$!destination}/$target" !! $!to-absolute);
+            return $p;
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:76]: `return Promise.kept(...)`

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            my $dest-p = $!destination($request);
+            if $dest-p ~~ Awaitable {
+                return $dest-p.then({ .status ~~ Kept ?? do { my $url = self!trim-url(.result);
+                                                              $!to ?? "$url/$target" !! "$url"
+                                                            } !! die 'Failed to get URL' });
+            } else {
+                my $p = Promise.new;
+                my $tmp = self!trim-url($dest-p);
+                $p.keep($!to ?? "$tmp/$target" !! $tmp);
+                return $p;
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:87]: `Promise.kept`

**Review by jnthn — COMMENTED** [review]
Think this is pretty close to ready to go in; a few tweaks.

File: lib/Cro/HTTP/Request.pm6 (hunk 2 above)
Comment (Altai-man) [reply lib/Cro/HTTP/Request.pm6:86]: IMO that's a bad idea.

To differentiate between HTTP versions we already show underlying HTTP/2 frames in the debug output, which is enough to judge. But when we show outright violation of the protocol in the debug output that's IMO much more confusing.

File: lib/Cro/HTTP/ReverseProxy.pm6 [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+class Cro::HTTP::ReverseProxy does Cro::Transform {
+    subset ReverseProxyLink where Str|Callable;
+
+    has ReverseProxyLink $.to;
+    has Str $.to-absolute;
+    has ReverseProxyLink $!destination;
+    has Cro::HTTP::Client $!client;
+    has Callable $.request;
+    has Callable $.response;
```
Comment (jnthn) [inline lib/Cro/HTTP/ReverseProxy.pm6:18]: This can also be expressed as `has &.request;`

**Review by jnthn — COMMENTED** [review]
Looks good to me overall. I wonder if we should have a test case for changing the request body or response body also?

File: t/http-reverse-proxy.t [hunk 11]
```diff
@@ -201,6 +205,28 @@ END {
...
+            is $response.body-text.result, 'body', 'Proxy set a new body in request';
+            $response.set-body('body updated');
+        });
+
+    my $proxy = Cro::HTTP::Server.new(port => HTTP_TEST_PORT_PROXY, application => $proxy-app);
+    $proxy.start;
+    LEAVE $proxy.stop;
+
+    my $c = Cro::HTTP::Client.new(base-uri => "http://localhost:{HTTP_TEST_PORT_PROXY}");
+    given await $c.get('/base-body') -> $resp {
```
Comment (jnthn) [inline t/http-reverse-proxy.t:225]: Should probably be a `POST` or `PUT` request if it's going to have a request body.

**Review by jnthn — APPROVED** [review]
